lib/bolt/node/winrm.rb

```diff
@@ -69,38 +69,71 @@ def shell_init
     $StdinInput = $Null
   )
 
-  $startInfo = New-Object System.Diagnostics.ProcessStartInfo($Path, $Arguments)
-  $startInfo.UseShellExecute = $false
-  $startInfo.WorkingDirectory = Split-Path -Parent (Get-Command $Path).Path
-  if ($StdinInput) { $startInfo.RedirectStandardInput = $true }
-  $startInfo.RedirectStandardOutput = $true
-  $startInfo.RedirectStandardError = $true
-
   try
   {
-    $process = [System.Diagnostics.Process]::Start($startInfo)
+    $startInfo = New-Object System.Diagnostics.ProcessStartInfo($Path, $Arguments)
+    $startInfo.UseShellExecute = $false
+    $startInfo.WorkingDirectory = Split-Path -Parent (Get-Command $Path).Path
+    $startInfo.CreateNoWindow = $true
+    if ($StdinInput) { $startInfo.RedirectStandardInput = $true }
+    $startInfo.RedirectStandardOutput = $true
+    $startInfo.RedirectStandardError = $true
+
+    $stdoutHandler = { if (-not ([String]::IsNullOrEmpty($EventArgs.Data))) { $Host.UI.WriteLine($EventArgs.Data) } }
+    $stderrHandler = { if (-not ([String]::IsNullOrEmpty($EventArgs.Data))) { $Host.UI.WriteErrorLine($EventArgs.Data) } }
+    $invocationId = [Guid]::NewGuid().ToString()
+
+    $process = New-Object System.Diagnostics.Process
+    $process.StartInfo = $startInfo
+    $process.EnableRaisingEvents = $true
+
+    # https://msdn.microsoft.com/en-us/library/system.diagnostics.process.standarderror(v=vs.110).aspx#Anchor_2
+    $stdoutEvent = Register-ObjectEvent -InputObject $process -EventName 'OutputDataReceived' -Action $stdoutHandler
+    $stderrEvent = Register-ObjectEvent -InputObject $process -EventName 'ErrorDataReceived' -Action $stderrHandler
+    $exitedEvent = Register-ObjectEvent -InputObject $process -EventName 'Exited' -SourceIdentifier $invocationId
+
+    $process.Start() | Out-Null
+
+    $process.BeginOutputReadLine()
+    $process.BeginErrorReadLine()
+
+    if ($StdinInput)
+    {
+      $process.StandardInput.WriteLine($StdinInput)
+      $process.StandardInput.Close()
+    }
+
+    # park current thread until the PS event is signaled upon process exit
+    # OR the timeout has elapsed
+    $waitResult = Wait-Event -SourceIdentifier $invocationId -Timeout $Timeout
+    if (! $process.HasExited)
```

Comment:
Another way of checking this is with `if ($waitResult.SourceIdentifier -eq $Null)`, but I think this is clearer